### PR TITLE
Added Golub Kahan Bidiagonalization with Householder Reflections

### PR DESCRIPTION
**Pull request by sudoWin**
#### Brief description of what is fixed or changed
Added Golub Kahan Bidiagonalization

#### Other comments

#18775

#### Release Notes

* matrices
  * Added Golub Kahan Bidiagonalization with Householder Reflections

### CODE REVIEW

**Comment by sudoWin** [comment]
I only added to the code of matrices.py, eigen.py, test_eigen.py, and durations.json, but for some unknown reason, tests in number theory are failing.

**Comment by sudoWin** [comment]
What should I do know? Can you please help a little @sylee957 .   The tests that fail have no relation with my PR

**Comment by sylee957** [comment]
I think that the test should be rerun as I deferred the issue in #18800

**Comment by sylee957** [comment]
Can the algorithm be modified as a decomposition method to return the transformation matrix `U` and `V`?

**Comment by sudoWin** [comment]
Yes, I will modify it and submit a PR soon.

File: sympy/matrices/eigen.py [hunk 1]
```diff
@@ -372,6 +373,76 @@ def _is_diagonalizable(M, reals_only=False, **kwargs):
...
+
+def _bidiagonalize(M, precision=None, upper=True):
+    """
+    Returns bidiagonalized form of the input Matrix.
+
+    Parameters
+    ==========
+
+    precision : int. Number of digits beside decimal point.
+                Default: None, which makes all the computation to be done symbollically
```
Comment (sylee957) [inline sympy/matrices/eigen.py]: I don't think that it's actually transparent inside how `precision` is supposed to work.
Like how this guarantees the precision
```
    if precision is not None:
        sig = N(sig)
```

And even I have some questions whether this keyword is needed or not. For example, you can first convert the matrix to floating point and then all the operations can be done with floating point arithmetics.

Comment (sudoWin) [reply sympy/matrices/eigen.py]: How should I convert the Matrix to  floating point? Like, is there any recommended or suggested way to convert a Sympy Matrix to floating-point.

And even if I convert the entire Matrix to floating-point, some error is always introduced of the order of 10^-16, what should I do regarding that?

Comment (sylee957) [reply sympy/matrices/eigen.py]: `Matrix.evalf` can have the logic. 
I think that a zero test using epsilon can only make sense for purely floating point matrix, so you may better handle that automatically after `all(isinstance(x, float) for x in M)`.

**Comment by sylee957** [comment]
I have a few suggestions

1. You should remove the `evalf` logic inside the code because sympy is not a numeric library and converting to floating points in priori can always be done from the user-side API.
For now, even if things getting hanging for symbolic computations, I'd rather view that as an open research problem than simply disallowing it in this way because of the above reason.
(But you should probably document that the computation can hang for symbolic matrices)
And I also think that it's more educationally meaningful to have a method to compute stuff like 2*2 matrix bidiagonalization with radicals.

2. You should correct the docs like `U * B * V.H` (I'd rather believe it's `U * B * V.H`) and polish the code about spacing between operators. And I think that there should be more tests than naive cases. (e.g. complex integer matrix or rectangular matrix, mutable and immutable matrix)

3. I'd suggest a dual API like `bidiagonalize` and `bidiagonal_decomposition`, and make `bidiagonalize` to return only `B` while the decomposition returns all the `U, B, V`.
I'd suggest this because a lot of the redundant computations for computing the `U, V` can be skipped for computing only the `B` matrix.

**Comment by sudoWin** [comment]
I have worked on your suggestions and here is the checklist,

- [x] Remove evalf logic

- [x] Added in the documentation, that computation can hang for big matrices

- [x] Corrected the docs. and also latex formatted them.(If you think, i should add more, do tell me)

- [x] Polished the code for spacing between operators

- [x] Complex Integer Support Added

- [x] Complex Integer Matrix Tests Added

- [x] Added Random Tests along with naive tests

- [ ] Added Immutable Matrix Test

- [ ] Added Rectangular Matrix Tests

- [x] Changed the API to have two functions, bidiagonalize, and bidiagonal_decomposition

**Comment by sudoWin** [comment]
For rectangular matrices, tests become too slow, so I had doubts on whether to add them or not. So should I add them.
For immutable matrices, I was unable to create a mutable matrix out of the immutable one, Is there a way to do so other than by using Matrix() ?

**Comment by sylee957** [comment]
If you have concerns about some slowdowns, I'd recommend adding them as floating point tests rather than radicals. I also anticipate much slowdowns for stuff larger than `2*2` matrix so we would only need `2*2` matrix tests for exact radicals and other tests can be made floating points.

There is `as_mutable` to convert the immutable matrix to its respectable type.

**Comment by sudoWin** [comment]
- [x] Added Immutable Matrix Suport
- [x] Added Immutable Matrix Test
- [x] Added Rectangular Matrix Tests

**Comment by sudoWin** [comment]
Is there anything else that needs to be done?

File: sympy/matrices/eigen.py [hunk 2]
```diff
@@ -372,6 +372,120 @@ def _is_diagonalizable(M, reals_only=False, **kwargs):
...
+    m = M.rows
+    n = M.cols
+    A = M.as_mutable()
+    for i in range(min(m, n)):
+        v, bet = _householder_vector(A[i:, i])
+        hh_mat = A.eye(m-i) - bet * v * v.H
+        A[i:, i:] = hh_mat * A[i:, i:]
+        if i + 1 <= n - 2:
+            v, bet = _householder_vector(A[i, i+1:].T)
+            hh_mat = A.eye(n-i-1)- bet * v * v.H
```
Comment (sylee957) [inline sympy/matrices/eigen.py]: Please fix the spacing between operators here

File: sympy/matrices/eigen.py [hunk 3]
```diff
@@ -372,6 +372,120 @@ def _is_diagonalizable(M, reals_only=False, **kwargs):
...
+    for i in range(min(m, n)):
+        v, bet = _householder_vector(A[i:, i])
+        hh_mat = A.eye(m - i) - bet * v * v.H
+        A[i:, i:] = hh_mat * A[i:, i:]
+        temp = A.eye(m)
+        temp[i:, i:] = hh_mat
+        U = U * temp
+        if i + 1 <= n - 2:
+            v, bet = _householder_vector( A[i, i+1:].T )
+            hh_mat = A.eye(n - i - 1)- bet * v * v.H
```
Comment (sylee957) [inline sympy/matrices/eigen.py]: Please fix the spacing between operators here

File: sympy/matrices/eigen.py [hunk 4]
```diff
@@ -372,6 +372,120 @@ def _is_diagonalizable(M, reals_only=False, **kwargs):
     return _is_diagonalizable_with_eigen(M, reals_only=reals_only)[0]
 
 
+#G&VL, Matrix Computations, Algo 5.4.2
```
Comment (sylee957) [inline sympy/matrices/eigen.py]: You'd also put the references in the docstrings of public docs.

File: sympy/matrices/eigen.py [hunk 5]
```diff
@@ -372,6 +372,120 @@ def _is_diagonalizable(M, reals_only=False, **kwargs):
...
+def _householder_vector(x):
+    if not x.cols == 1:
+        raise ValueError("Input must be a column matrix")
+    v = x.copy()
+    v_plus = x.copy()
+    v_minus = x.copy()
+    q = x[0, 0]/abs(x[0, 0])
+    norm_x = x.norm()
+    v_plus[0, 0] = x[0, 0] + q * norm_x
+    v_minus[0, 0] = x[0,0] - q * norm_x
```
Comment (sylee957) [inline sympy/matrices/eigen.py]: Please fix the spacing of `x[0,0]` here

File: sympy/matrices/matrices.py [hunk 6]
```diff
@@ -378,6 +380,18 @@ def diagonalize(self, reals_only=False, sort=False, normalize=False):
...
+    def _bidiag_hholder(self):
+        return _eval_bidiag_hholder(self)
+
+    def bidiagonalize(self, upper=True):
+        return _bidiagonalize(self, upper=upper)
+
+    def bidiagonal_decomposition(self,upper=True):
+        return _bidiagonal_decomposition(self,upper=upper)
+
+    def bidiagonal_decmp_hholder(self):
```
Comment (sylee957) [inline sympy/matrices/matrices.py]: Why is this needed?

Comment (sudoWin) [reply sympy/matrices/matrices.py]: No particular reason, but while reading Sympy code earlier, I had seen multiple functions, like

```python3
_eval_det_lu,
_eval_det_bareiss
_eval_det_berkowitz
```
along with
```python
det
```

so I did the same way.
Should I remove
```python
bidiagonal_decmp_hholder
```

Comment (sylee957) [reply sympy/matrices/matrices.py]: `_eval_det` stuff are private methods while `bidiagonal_decmp_hholder` is a public method. So we should remove the redundant public method.

Comment (sudoWin) [reply sympy/matrices/matrices.py]: Okay, removed `bidiagonal_decmp_hholder`

File: sympy/matrices/eigen.py [hunk 7]
```diff
@@ -372,6 +372,132 @@ def _is_diagonalizable(M, reals_only=False, **kwargs):
...
+        if i + 1 <= n - 2:
+            v, bet = _householder_vector(A[i, i+1:].T)
+            hh_mat = A.eye(n - i - 1) - bet * v * v.H
+            A[i:, i+1:] = A[i:, i+1:] * hh_mat
+    return A
+
+
+def _bidiagonal_decomposition(M, upper=True):
+    """
+    Returns (U,B,V.H)
```
Comment (sylee957) [inline sympy/matrices/eigen.py]: Should it be `U, B, V`?

Comment (sudoWin) [reply sympy/matrices/eigen.py]: I rechecked this research paper 
https://www.researchgate.net/publication/220949174_Matrix_Bidiagonalization_on_the_Trident_Processor
and I am positive that it is `U, B, V.T` for real and `U, B, V.H` but `V.T = V.H` for real matrices, so we can say that it is `U, B, V.H`

Comment (sudoWin) [reply sympy/matrices/eigen.py]: There isn't any mention of it in Matrix Computations, by G&VL, and nearly nowhere else,(I couldn't find one).

File: sympy/matrices/eigen.py [hunk 8]
```diff
@@ -372,6 +372,132 @@ def _is_diagonalizable(M, reals_only=False, **kwargs):
...
+def _householder_vector(x):
+    if not x.cols == 1:
+        raise ValueError("Input must be a column matrix")
+    v = x.copy()
+    v_plus = x.copy()
+    v_minus = x.copy()
+    q = x[0, 0] / abs(x[0, 0])
+    norm_x = x.norm()
+    v_plus[0, 0] = x[0, 0] + q * norm_x
+    v_minus[0, 0] = x[0, 0] - q * norm_x
```
Comment (sylee957) [inline sympy/matrices/eigen.py]: I wonder where this method is from
I think that it's from https://github.com/vslobody/Householder-Bidiagonalization but it looks different than Golub & Van Loan's book. 
I know that the householder vector can be chosen in some arbitrary way. So it can be a trivial problem if this is correct.

File: sympy/matrices/eigen.py (hunk 7 above)
Comment (sylee957) [reply sympy/matrices/eigen.py]: Okay, I see in this version it already returns the transposed form

File: sympy/matrices/eigen.py (hunk 8 above)
Comment (sudoWin) [reply sympy/matrices/eigen.py]: The original method which I had written was from  Golub & Van Loan's book, with some tweaks to remove redundant operations, but later I found out that the said method didn't work for complex matrices, so I looked for another method and found this one.

**Comment by oscarbenjamin** [comment]
I think this can be improved.

The test code here is very confusing because the asserts are a long way away from the calculations. It would be better to put them right next to them like:
```
A, B, C = M.bidiagonalize_components()
assert A * B * C == M
```
There can sometimes be reasons for using random tests but I don't see the merit here because:
- The randomised tests are only for 2x2 matrices in which case bidiagonalisation is fairly trivial.
- Basic cases are not properly tested

We also normally try to avoid using generic simplify in testing or library code because usually a specific simplification function is better. Why is simplify needed? I guess from an example that cancel is sufficient:
```
In [43]: M = Matrix([[1, 2], [3, 4]])                                                                                                                         

In [44]: A, B, C = M.bidiagonal_decomposition()                                                                                                               

In [45]: A                                                                                                                                                    
Out[45]: 
⎡               2                            -6                 ⎤
⎢    1 - ───────────────         ───────────────────────────    ⎥
⎢                 9                        ⎛         9     ⎞    ⎥
⎢        1 + ───────────         (1 - √10)⋅⎜1 + ───────────⎟    ⎥
⎢                      2                   ⎜              2⎟    ⎥
⎢            (-1 + √10)                    ⎝    (-1 + √10) ⎠    ⎥
⎢                                                               ⎥
⎢            -6                              18                 ⎥
⎢───────────────────────────  - ──────────────────────────── + 1⎥
⎢          ⎛         9     ⎞             2 ⎛         9     ⎞    ⎥
⎢(1 - √10)⋅⎜1 + ───────────⎟    (1 - √10) ⋅⎜1 + ───────────⎟    ⎥
⎢          ⎜              2⎟               ⎜              2⎟    ⎥
⎣          ⎝    (-1 + √10) ⎠               ⎝    (-1 + √10) ⎠    ⎦

In [52]: A.applyfunc(cancel)                                                                                                                                  
Out[52]: 
⎡ -(-1 + √10)     -(-33 + 6⋅√10) ⎤
⎢ ────────────    ───────────────⎥
⎢  -10 + √10        -20 + 11⋅√10 ⎥
⎢                                ⎥
⎢-(-33 + 6⋅√10)     -31 + 13⋅√10 ⎥
⎢───────────────   ───────────── ⎥
⎣  -20 + 11⋅√10    -130 + 31⋅√10 ⎦

In [53]: A.applyfunc(cancel).applyfunc(radsimp)                                                                                                               
Out[53]: 
⎡ √10   3⋅√10⎤
⎢ ───   ─────⎥
⎢  10     10 ⎥
⎢            ⎥
⎢3⋅√10  -√10 ⎥
⎢─────  ─────⎥
⎣  10     10 ⎦

In [47]: cancel(cancel(A)*cancel(B)*C) == M                                                                                                                   
Out[47]: True
```
Maybe it's worth calling cancel on the outputs by default if this kind of simplification is usually beneficial.

I just tried this example which seems to hang:
```
In [1]: M = Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])                                                                                                         

In [2]: M.bidiagonalize()                                                                                                                                     
^C---------------------------------------------------------------------------
KeyError 
```

This also fails for simple symbolic examples:
```julia
In [57]: x = Symbol('x', real=True)                                                                                                                                      

In [58]: Matrix([[1, x], [x, 1]]).bidiagonalize()                                                                                                             
---------------------------------------------------------------------------
TypeError: cannot determine truth value of Relational
```

**Comment by sudoWin** [comment]
Thanks @oscarbenjamin for suggesting multiple changes,

1. I added a comment that described what the variables were storing, I will submit a new PR and move the asserts right next to calculations.
2. Could you please suggest which basic cases I should add? I would be more than happy to add them.
3. I added random tests because I thought it might be beneficial, I can remove it if you say so.
What I thought while adding them was, that every time Testing is done, the code gets checked against a totally new Matrix, and I thought that would be better.
4.The reason behind using simplify is to simplify it as much as possible so that final value should be an integer, which it MUST be. 

In the above example itself, one has to apply

cancel
radsimp
cancel
product
cancel

Further, for assert to not raise an Assertion Error, the matrices have to be exactly equal,
and I wasn't sure after how many cancellations and simplifications, the result would be exactly equal, so that's why I used simplify()

5.We can use cancel on the outputs by default, but again, there's still a question that
After how many cancel and radsimp would the output be so simplified that the user doesn't need to invoke simplify.

6.Examples of 3x3 or larger will hang for symbolic computation, and if the input is converted to a floating point, well it can easily handle upto 25x25. I am looking on how to do this symbollically and with better algorithms,(and I have even found one and I am in process of implementing it).Also 
As sylee957 said

> You should remove the evalf logic inside the code because sympy is not a numeric library and converting to floating points in priori can always be done from the user-side API.
For now, even if things getting hanging for symbolic computations, I'd rather view that as an open research problem than simply disallowing it in this way because of the above reason.
(But you should probably document that the computation can hang for symbolic matrices)
And I also think that it's more educationally meaningful to have a method to compute stuff like 2*2 matrix bidiagonalization with radicals.
I will try speeding it up.

7.Regarding the usage of symbols, I dont have a solution, but I would try working to solve that one.

**Comment by sudoWin** [comment]
I tried doing the 
```python
In [57]: x = Symbol('x', real=True)                                                                                                                                      

In [58]: Matrix([[1, x], [x, 1]]).bidiagonalize()     
```
and the entire problem breaks down to:
```python
In [51]: x = Symbol('x', real=True)                                                                                                                                                                         

In [52]: abs(1 + sqrt(1 + x**2)) >= abs(1 - sqrt(1 + x**2))                                                                                                                                                 
Out[52]: 
   ________       │   ________    │
  ╱  2            │  ╱  2         │
╲╱  x  + 1  + 1 ≥ │╲╱  x  + 1  - 1│

```
now this above expression is always `True`. Is there any function in Sympy that can solve this inequality? If there is one, I can try using that in the householder vector. Meanwhile, I will keep looking for a workaround.

**Comment by oscarbenjamin** [comment]
Is it fundamentally necessary to know which of the two expressions is bigger in order to make the algorithm work?

This is important because it will be the ultimate limitation in the algorithm for the symbolic case.

**Comment by sudoWin** [comment]
I am not sure, but after a lot of random tests and some pen and paperwork, I found out that the else block is responsible for flipping the sign of the diagonal element corresponding to the column being operated on, to make all diagonal values positive.
```python
if v_plus.norm() <= v_minus.norm():
   v = v_plus
else:
   v = v_minus
```
 and if we remove that, it would still work fine, the only thing is that signs of some values, get flipped. Technically it should be correct because the householder matrix just reflects about a plane/hyperplane and it seems totally valid as the definition of a bidiagonal matrix nowhere states that the diagonal elements must be positive. However, I couldn't find any official source or research paper stating this, so I am not 100% sure. What's your opinion on this, sir?

**Comment by sylee957** [comment]
I think that `v_plus` and `v_minus` can be chosen arbitrarily, but the first zero testing may not be skipped. But I think that the check can be relaxed 

I'm also thinking that just using the direct formula 

image 

can be better than using the Golub's construction

Because it gives a worse result for symbolic construction. (Latter is Golub's construction)

image
image

**Comment by sudoWin** [comment]
I realized and agree that v_plus and v_minus can be chosen arbitrarily, however I also think that both of them can be eliminated and we can also eliminate the check.
I have currently submitted a PR, which does the above and since it was the main limitation, this also add basic support for symbolic matrices.
Currently, the new PR uses Golub's Construction, however I am currently looking whether the direct formula can be implemented with better and improved results.
